### PR TITLE
Test suite - Complete CMake restructure and CI issue fixes

### PR DESCRIPTION
**Pull request by r-abishek**
1. Complete restructure and overhaul of CMake for RPP test suite to bring clarity in the dependencies for the 4 tensor test suite types (image/voxel/audio/misc)
2. Fixes for all CI issues - [Issue] - #454 #449 #427 
3. Add fix for test suite hip build - add iomanip include
4. Ensure all test suite dependencies are checked in CMakeLists
5. Ensure all test suite dependencies are listed correctly in README
6. Support to document and clarify 4 test sets being run (image/voxel/audio/misc)
7. Support to ensure other test sets work if dependencies for one test set are not present on the machine
8. Restructure test suite headers

### CODE REVIEW

**Comment by kiritigowda** [comment]
@r-abishek can you add the fix for Issue #454 here?

**Comment by r-abishek** [comment]
@kiritigowda Added iomanip PR updates here - https://github.com/r-abishek/rpp/pull/344

**Comment by kiritigowda** [comment]
@r-abishek any updates on this PR? Need the fix to solve test failures

**Comment by kiritigowda** [comment]
@rrawther - this fix is required to pass CI -- issue mentioned -- #454

**Comment by rrawther** [comment]
@r-abishek : We need to merge this PR asap as CI is failing on bunch of PRs. Is it still WIP?

File: utilities/test_suite/HIP/CMakeLists.txt [hunk 1]
```diff
@@ -67,6 +67,7 @@ find_library(libsnd_LIBS
 find_package(OpenMP REQUIRED)
 set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} ${OpenMP_C_FLAGS}")
 set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${OpenMP_CXX_FLAGS}")
+set(LINK_LIBRARY_LIST ${LINK_LIBRARY_LIST} OpenMP::OpenMP_CXX)
```
Comment (rrawther) [inline utilities/test_suite/HIP/CMakeLists.txt:70]: why OpenMP::OpenMP_CXX is required in link library list?

**Review by rrawther — CHANGES_REQUESTED** [review]
Added a comment

**Comment by r-abishek** [comment]
@kiritigowda @rrawther Please see screenshots below with lsb_release -a command to show Ubuntu passing on 22.04 and 20.04:

Passing on Ubuntu 22.04:
image

Passing on Ubuntu 20.04:
image

**Comment by r-abishek** [comment]
@kiritigowda @rrawther Current fails on CI are just since the right libs are not installed on the CI machine for both SLES and RHEL:

On RHEL9:
image

On SLES15:
image

**Comment by kiritigowda** [comment]
@r-abishek - we are installing all the deps for CI. The failure on RHEL 8 can be verified with this fix.

**Review by rrawther — DISMISSED** [review]

**Review by kiritigowda — APPROVED** [review]

**Comment by kiritigowda** [comment]
@r-abishek - errors from this PR - http://math-ci.amd.com/job/mainline/job/precheckin/job/rpp/job/PR-453/5/pipeline-console/log?nodeId=282

```
: In file included from /opt/rocm-6.3.0-14898/share/rpp/test/HOST/Tensor_misc_host.cpp:25:
1: /opt/rocm-6.3.0-14898/share/rpp/test/HOST/../rpp_test_suite_misc.h:97:39: error: implicit instantiation of undefined template 'std::array<unsigned int, 4>'
1:                 std::array<Rpp32u, 4> roi = {0, 0, 100, 100};
1:                                       ^
1: /usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/tuple:1595:45: note: template is declared here
1:   template<typename _Tp, size_t _Nm> struct array;
1:                                             ^
1: In file included from /opt/rocm-6.3.0-14898/share/rpp/test/HOST/Tensor_misc_host.cpp:25:
1: /opt/rocm-6.3.0-14898/share/rpp/test/HOST/../rpp_test_suite_misc.h:104:39: error: implicit instantiation of undefined template 'std::array<unsigned int, 6>'
1:                 std::array<Rpp32u, 6> roi = {0, 0, 0, 50, 50, 8};
1:                                       ^
1: /usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/tuple:1595:45: note: template is declared here
1:   template<typename _Tp, size_t _Nm> struct array;
1:                                             ^
1: In file included from /opt/rocm-6.3.0-14898/share/rpp/test/HOST/Tensor_misc_host.cpp:25:
1: /opt/rocm-6.3.0-14898/share/rpp/test/HOST/../rpp_test_suite_misc.h:118:39: error: implicit instantiation of undefined template 'std::array<unsigned int, 4>'
1:                 std::array<Rpp32u, 4> roi = {0, 0, 1920, 1080};
1:                                       ^
1: /usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/tuple:1595:45: note: template is declared here
1:   template<typename _Tp, size_t _Nm> struct array;
1:                                             ^
1: In file included from /opt/rocm-6.3.0-14898/share/rpp/test/HOST/Tensor_misc_host.cpp:25:
1: /opt/rocm-6.3.0-14898/share/rpp/test/HOST/../rpp_test_suite_misc.h:125:39: error: implicit instantiation of undefined template 'std::array<unsigned int, 6>'
1:                 std::array<Rpp32u, 6> roi = {0, 0, 0, 1920, 1080, 3};
1:                                       ^
1: /usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/tuple:1595:45: note: template is declared here
1:   template<typename _Tp, size_t _Nm> struct array;
1:                                             ^
1: In file included from /opt/rocm-6.3.0-14898/share/rpp/test/HOST/Tensor_misc_host.cpp:25:
1: /opt/rocm-6.3.0-14898/share/rpp/test/HOST/../rpp_test_suite_misc.h:132:39: error: implicit instantiation of undefined template 'std::array<unsigned int, 8>'
1:                 std::array<Rpp32u, 8> roi = {0, 0, 0, 0, 1, 128, 128, 128};
1:                                       ^
1: /usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/tuple:1595:45: note: template is declared here
1:   template<typename _Tp, size_t _Nm> struct array;
1:                                             ^
1: 5 errors generated.
1: make[3]: *** [CMakeFiles/Tensor_misc_host.dir/build.make:76: CMakeFiles/Tensor_misc_host.dir/Tensor_misc_host.cpp.o] Error 1
1: make[3]: Leaving directory '/var/jenkins_home/workspace/mainline_precheckin_rpp_PR-453/govcksC4S/rpp/build/release/utilities/test_suite/build'
1: make[2]: *** [CMakeFiles/Makefile2:119: CMakeFiles/Tensor_misc_host.dir/all] Error 2
1: make[2]: *** Waiting for unfinished jobs....
```

Check full report - http://math-ci.amd.com/job/mainline/job/precheckin/job/rpp/job/PR-453/5/pipeline-console/?selected-node=279

**Comment by r-abishek** [comment]
@kiritigowda All Jenkins CI passing now for this PR - Ubuntu/SLES/RHEL.
Strict checks introduced in CMake for all packages necessary to build and run each of the 4 test sets.
PR description changed to point to all issues it addresses.

**Comment by r-abishek** [comment]
@kiritigowda @rrawther Last commit is really taking too long to run the CI.

Please see CI passing on the second last commit.
image

**Comment by kiritigowda** [comment]
Merge pending - http://math-ci.amd.com/job/mainline/job/precheckin/job/rpp/job/PR-453/16/pipeline-graph/

**Comment by r-abishek** [comment]
> Merge pending - http://math-ci.amd.com/job/mainline/job/precheckin/job/rpp/job/PR-453/16/pipeline-graph/

@kiritigowda The Ubuntu22+gfx90a may not finish since:
image

**Comment by r-abishek** [comment]
@kiritigowda All 13 CI passing now

**Pull request merged**